storm-client/src/jvm/org/apache/storm/daemon/Task.java

```diff
@@ -112,15 +112,15 @@ public Task(Executor executor, Integer taskId) throws IOException {
     public List<Integer> getOutgoingTasks(Integer outTaskId, String stream, List<Object> values) {
         if (debug) {
             LOG.info("Emitting direct: {}; {} {} {} ", outTaskId, componentId, stream, values);
-        }
```

Comment:
I think grouping check should happen even if debug is not enabled.

Reply:
That was part of an older PR. From what I recall ... on profiling I had noticed that the grouping check was expensive in the critical path due to the fact that it needed lookups in three (now down to 2) hashmaps :   streamComponentToGrouper  &  componentGrouping. Since neither were keyed on Integer, the CustomIndexArray style optimization was not possible.